File: lcp.py
def longestCommonPrefix(strs):
    if not strs:
        return ""
    
    # Take the first string as a reference
    prefix = strs[0]
    
    for i in range(1, len(strs)):
        # Keep reducing the prefix until it matches the current string
        while strs[i].find(prefix) != 0:
            prefix = prefix[:-1]
            if not prefix:
                return ""
    
    return prefix

# Test the function
strs = ["flower", "flow", "flight"]
print(longestCommonPrefix(strs))  # Output: "fl"

# Additional test cases
print(longestCommonPrefix(["dog", "racecar", "car"]))  # Output: ""
print(longestCommonPrefix(["interspecies", "interstellar", "interstate"]))  # Output: "inters"
print(longestCommonPrefix(["throne", "throne"]))  # Output: "throne"
print(longestCommonPrefix(["", "b"]))  # Output: ""
print(longestCommonPrefix([]))  # Output: ""

File: lcp.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* longestCommonPrefix(char** strs, int strsSize) {
    if (strsSize == 0) return "";
    
    int i, j;
    int min_len = strlen(strs[0]);
    
    // Find the minimum length string
    for (i = 1; i < strsSize; i++) {
        int len = strlen(strs[i]);
        if (len < min_len) min_len = len;
    }
    
    // Allocate memory for the result
    char* result = (char*)malloc((min_len + 1) * sizeof(char));
    
    // Compare characters
    for (i = 0; i < min_len; i++) {
        char c = strs[0][i];
        for (j = 1; j < strsSize; j++) {
            if (strs[j][i] != c) {
                result[i] = '\0';
                return result;
            }
        }
        result[i] = c;
    }
    
    result[i] = '\0';
    return result;
}

int main() {
    char* strs[] = {"flower", "flow", "flight"};
    int strsSize = sizeof(strs) / sizeof(strs[0]);
    
    char* result = longestCommonPrefix(strs, strsSize);
    printf("Longest common prefix: %s\n", result);
    
    free(result);
    return 0;
}
